tuxedo-core/src/constraint_checker.rs

```diff
@@ -7,11 +7,111 @@ use sp_std::{fmt::Debug, vec::Vec};
+    pub priority: TransactionPriority,
+    /// An intermediate value that should be passed to an accumulator that track transient intra-block data.
+    pub accumulator_value: ValueType,
+}
+
+impl<ValueType> ConstraintCheckingSuccess<ValueType> {
+    pub fn transform<NewValueType: From<ValueType>>(
+        self,
+    ) -> ConstraintCheckingSuccess<NewValueType> {
+        ConstraintCheckingSuccess::<NewValueType> {
+            priority: self.priority,
+            accumulator_value: self.accumulator_value.into(),
+        }
+    }
+}
+
+impl<ValueType> From<TransactionPriority> for ConstraintCheckingSuccess<ValueType>
+where
+    ValueType: Default,
+{
+    fn from(priority: TransactionPriority) -> ConstraintCheckingSuccess<ValueType> {
+        ConstraintCheckingSuccess::<ValueType> {
+            priority,
+            accumulator_value: ValueType::default(),
+        }
+    }
+}
+
+impl<ValueType> From<(TransactionPriority, ValueType)> for ConstraintCheckingSuccess<ValueType> {
+    fn from(t: (TransactionPriority, ValueType)) -> ConstraintCheckingSuccess<ValueType> {
+        ConstraintCheckingSuccess::<ValueType> {
+            priority: t.0,
+            accumulator_value: t.1,
+        }
+    }
+}
+
+/// An accumulator allows a Tuxedo piece to do some internal bookkeeping during the course
+/// of a single block. The Bookkeeping must be done through this accumulator-like interface
+/// where each transaction yields some intermediate value that is them folded into the accumulator
+/// via som logic that is encapsulated in the impelementation.
+///
+/// Many Tuxedo pieces will not need accumulators. In such a case, the constraint checker should
+/// simply use () as the accumulator type.
+///
+/// Some typical usecases for an accumulator would be making sure that the block author reward
+/// does not exceed the block's transaction fees. Or making sure that a timestamp transaction occurs
+/// at most once in a single block.
+///
+/// The accumulator is reset automatically at the end of every block. It is not possible to store
+/// data here for use in subsequent blocks.
+pub trait Accumulator {
+    /// The type that is given and also the type of the accumulation result.
+    /// I realize that the most general accumulator swill use two different types for those,
+    /// but let's do that iff we ever need it. I probably will need to so I can do a simple counter.
+    type ValueType: Debug + Encode + Decode;
+
+    /// The accumulator value that should be used to start a fresh accumulation
+    /// at the beginning of each new block.
+    ///
+    /// This is a function and takes a value, as opposed to being a constant for an important reason.
+    /// Aggregate runtimes made from multiple pieces will need to give a different initial value depending
+    /// which of the constituent constraint checkers is being called.
+    fn initial_value(_: Self::ValueType) -> Self::ValueType;
```

Comment:
Will this just always be the `Default` for the `ValueType` associated type? Should `ValueType` just instead implement the `Default` trait?

Reply:
I thought about that same thing. All the examples of accumulators that I thought of can indeed be implemented using the default value as the initial value.

I ended up going with a separate method for these reasons:
1. Semantics. While it seems the initial value of the accumulator will typically be the default, there is no fundamental reason it must be this way. Someone may write an intuitive impl of `Default` for that purpose.
2. To mirror https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold
3. The main one: For aggregate runtimes. When the constraint checker is composed with the macro from individual constraint checkers, it needs to do an actual match and return a different initial value depending on which piece is being called. Default does allow taking a param and returning a different result based on that param.